contracts/P256Signer.sol

```diff
@@ -51,20 +51,38 @@ contract P256Signer {
     /// @param _hash The hash of the data signed
     /// @param _signature The signature
     /// @return The EIP-1271 magic value
-    function isValidSignature(bytes memory _hash, bytes memory _signature) public view returns (bytes4) {
+    function isValidSignature(bytes memory _hash, bytes calldata _signature) public view returns (bytes4) {
         _validate(_hash, _signature);
         return OLD_EIP1271_MAGICVALUE;
     }
 
+    struct EncodedSignature {
+        bytes authenticatorData;
+        bytes clientData;
+        uint256 challengeOffset;
+        uint256[2] rs;
+    }
+
     /// @notice Validates the signature
     /// @param data The data signed
     /// @param _signature The signature
-    function _validate(bytes memory data, bytes memory _signature) private view {
+    function _validate(bytes memory data, bytes calldata _signature) private view {
         bytes32 _hash = keccak256(data);
-        (bytes memory authenticatorData, bytes memory clientData, uint256 challengeOffset, uint256[2] memory rs) =
-            abi.decode(_signature, (bytes, bytes, uint256, uint256[2]));
-
-        bool valid = WrapperFCLWebAuthn.checkSignature(authenticatorData, 0x01, clientData, _hash, challengeOffset, rs, [x, y]);
+        EncodedSignature calldata parsedSignature;
+        assembly {
+            parsedSignature := _signature.offset
```

Comment:
Do you know if this generates code for bounds checks on the dynamic `calldata` fields from the structure?

As in, what happens if you pass in malformed `calldata` to this method?

Reply:
So, a fuzz test verifies that the method fails in case of an invalid signature (random bytes array). I commented on it above because I had to adjust the expected revert message. The trace suggested that it failed with `panic: array out-of-bounds access (0x32)`, so I bet it generates code for bound checks. Will verify.

Reply:
Had crossed my mind. I guess it works while everything still points within (calldata is large enough)?

Reply:
So this is the IR of the above code, including the call to checkSignature:
```
                        let var__hash_97 := expr_100
                        /// @src 0:2485:2526  "EncodedSignature calldata parsedSignature"
                        let var_parsedSignature_104_offset
                        let zero_t_struct$_EncodedSignature_$88_calldata_ptr_12_offset := zero_value_for_split_t_struct$_EncodedSignature_$88_calldata_ptr()
                        var_parsedSignature_104_offset := zero_t_struct$_EncodedSignature_$88_calldata_ptr_12_offset
                        /// @src 0:2536:2605  "assembly {..."
                        {
                            var_parsedSignature_104_offset := var__signature_93_offset
                        }
                        /// @src 0:2628:2640  "FCL_WebAuthn"
                        let expr_109_address := linkersymbol("lib/FreshCryptoLib/solidity/src/FCL_Webauthn.sol:FCL_WebAuthn")
                        /// @src 0:2669:2684  "parsedSignature"
                        let _13_offset := var_parsedSignature_104_offset
                        let expr_111_offset := _13_offset
                        /// @src 0:2669:2702  "parsedSignature.authenticatorData"
                        let _14 := add(expr_111_offset, 0)
                        let expr_112_offset, expr_112_length := access_calldata_tail_t_bytes_calldata_ptr(expr_111_offset, _14)
                        /// @src 0:2716:2720  "0x01"
                        let expr_113 := 0x01
                        /// @src 0:2734:2749  "parsedSignature"
                        let _15_offset := var_parsedSignature_104_offset
                        let expr_114_offset := _15_offset
                        /// @src 0:2734:2760  "parsedSignature.clientData"
                        let _16 := add(expr_114_offset, 32)
                        let expr_115_offset, expr_115_length := access_calldata_tail_t_bytes_calldata_ptr(expr_114_offset, _16)
                        /// @src 0:2774:2779  "_hash"
                        let _17 := var__hash_97
                        let expr_116 := _17
                        /// @src 0:2793:2808  "parsedSignature"
                        let _18_offset := var_parsedSignature_104_offset
                        let expr_117_offset := _18_offset
                        /// @src 0:2793:2824  "parsedSignature.challengeOffset"
                        let _19 := add(expr_117_offset, 64)
                        let expr_118 := read_from_calldatat_uint256(_19)
                        /// @src 0:2838:2853  "parsedSignature"
                        let _20_offset := var_parsedSignature_104_offset
                        let expr_119_offset := _20_offset
                        /// @src 0:2838:2856  "parsedSignature.rs"
                        let _21 := add(expr_119_offset, 96)
                        let expr_120_offset := _21
                        /// @src 0:2870:2871  "x"
                        let _22 := read_from_storage_split_offset_0_t_uint256(0x01)
                        let expr_121 := _22
                        /// @src 0:2885:2886  "y"
                        let _23 := read_from_storage_split_offset_0_t_uint256(0x02)
                        let expr_122 := _23
                        /// @src 0:2628:2896  "FCL_WebAuthn.checkSignature(..."
                        let _24 := convert_t_rational_1_by_1_to_t_bytes1(expr_113)
                        let expr_123 := fun_checkSignature_434(expr_112_offset, expr_112_length, _24, expr_115_offset, expr_115_length, expr_116, expr_118, expr_120_offset, expr_121, expr_122)
                        ```
So, if I'm reading this right, it calculates the offsets (and length for dynamic types) based on the struct layout and passes them to check the checkSIgnatures function. I think @cristovaoth is right about everything working if the calldata is large enough.

Reply:
It looks like for values, it uses `read_from_calldatat_uint256` to read values from `calldata` and for dynamic bytes, it uses `access_calldata_tail_t_bytes_calldata_ptr`. What do those YUL function look like?

Reply:
```
                    function access_calldata_tail_t_bytes_calldata_ptr(base_ref, ptr_to_tail) -> addr, length {
                        let rel_offset_of_tail := calldataload(ptr_to_tail)
                        if iszero(slt(rel_offset_of_tail, sub(sub(calldatasize(), base_ref), sub(0x20, 1)))) { revert_error_356d538aaf70fba12156cc466564b792649f8f3befb07b071c91142253e175ad() }
                        addr := add(base_ref, rel_offset_of_tail)

                        length := calldataload(addr)
                        if gt(length, 0xffffffffffffffff) { revert_error_1e55d03107e9c4f1b5e21c76a16fba166a461117ab153bcce65e6a4ea8e5fc8a() }
                        addr := add(addr, 32)
                        if sgt(addr, sub(calldatasize(), mul(length, 0x01))) { revert_error_977805620ff29572292dee35f70b0f3f3f73d3fdd0e9f4d7a901c2e43ab18a2e() }

                    }
```
and
```
                    function read_from_calldatat_uint256(ptr) -> returnValue {

                        let value := calldataload(ptr)
                        validator_revert_t_uint256(value)

                        returnValue :=

                        value

                    }
                    
                   function validator_revert_t_uint256(value) {
                        if iszero(eq(value, cleanup_t_uint256(value))) { revert(0, 0) }
                    }
                    
                   function cleanup_t_uint256(value) -> cleaned {
                        cleaned := value
                    }
```